client/protocols/analytics.py

```diff
@@ -89,5 +96,87 @@ def replaced(self, contents):
         return True
 
+    @classmethod
+    def read_history(cls):
+        history = {'questions': {}, 'all_attempts': 0}
+        try:
+            with open(cls.ANALYTICS_FILE, 'rb') as fp:
+                history = pickle.load(fp)
+            log.info('Loaded %d history from %s',
+                     len(history), cls.ANALYTICS_FILE)
+        except (IOError, EOFError) as e:
+            log.info('Error reading from ' + cls.ANALYTICS_FILE + \
+                     ', assume no history')
+        return history
+
+    def log_run(self, messages):
+        """Record this run of the autograder to a local file.
+        """
+        history = self.read_history()
+        history['all_attempts'] += 1
+        analytics = messages['analytics']
+        questions = analytics.get('question', [])
+        grading = 'grading' in messages and messages['grading']
+
+        # Attempt to figure out what question is being worked on
+        if not questions and grading:
+            failed = first_failed_test(self.assignment.specified_tests,
+                                       grading)
+            logging.info('First failed test: %s', failed)
+            if failed:
+                questions = [failed]
+            history['question'] = questions
+
+            # Update earlier question correctness status
+            for saved_q, details in history['questions'].items():
+                finished = details['solved']
+                if not finished and saved_q in grading:
+                    score = grading[saved_q]
+                    details['solved'] = is_correct(score)
+        else:
+            history['question'] = questions
+
+        for question in questions:
+            detail = history['questions']
+            if grading and question in grading:
+                score = is_correct(grading[question])
+            else:
+                score = 'Unknown'
+
+            if question in history['questions']:
+                q_info = detail[question]
+                if grading and question in grading:
+                    if q_info['solved'] != True:
+                        q_info['solved'] = score
+                    else:
+                        continue # Already solved. Do not change total
+                q_info['attempts'] += 1
+            else:
+                detail[question] = {
+                    'attempts': 1,
+                    'solved': score
+                }
+            logging.info('Attempt %d for Question %s : %r',
+                         history['questions'], question, score)
+
+        with open(self.ANALYTICS_FILE, 'wb') as f:
+            log.info('Saving history to %s', self.ANALYTICS_FILE)
+            pickle.dump(history, f)
+            os.fsync(f)
+
+        messages['analytics']['history'] = history
+
+def is_correct(score):
+    """Given a score from the grading protocol, see if no failed cases
+    and no locked cases.
+    """
+    return sum(score.values()) == score['passed']
```

Comment:
I'd have thought a score was just a single number, so I'm surprised to see that it's a dictionary. Please add a description in the docstring of what score's datatype is and what info it contains.